src/util/expr_iterator.h

```diff
@@ -0,0 +1,229 @@
+class const_depth_iteratort;
+// Depth first-search iterator with duplicate avoidance and a .mutate method
+// Slower iteration than naive version. Very expensive copy.
+class unique_depth_iteratort;
+// Depth first-search iterator with duplicate avoidance
+// Slower iteration than naive version. Very expensive copy.
+class const_unique_depth_iteratort;
+
+// Depth first search iterator - iterates over supplied expression
+// and all its operands recursively.
+// Base class using CRTP
+
+template<typename depth_iterator_t>
+class depth_iterator_baset:
+public std::iterator<std::forward_iterator_tag, const exprt, void>
+{
+public:
+  bool operator==(const depth_iterator_t &other) const
+  {
+    return m_stack.size()==other.m_stack.size() &&
+           std::equal(m_stack.begin(),
+                      m_stack.end(),
+                      other.m_stack.begin());
+  }
+  bool operator!=(const depth_iterator_t &other) const
+  {
+    return !(*this==other);
+  }
+  depth_iterator_t &operator++()
+  {
+    PRECONDITION(!m_stack.empty());
+    while(true)
+    {
+      if(m_stack.back().it==m_stack.back().end)
+      {
+        m_stack.pop_back();
+        if(m_stack.empty())
+        {
+          break;
+        }
+      }
+      else if(this->downcast().push_expr(*m_stack.back().it))
+      {
+        break;
+      }
+      m_stack.back().it++;
+    }
+    return this->downcast();
+  }
+  const exprt &operator*() const
+  {
+    return m_stack.back().expr.get();
+  }
+  const exprt *operator->() const
+  {
+    return &**this;
+  }
+protected:
+  // Create end iterator
+  depth_iterator_baset() { }
+  // Create begin iterator, starting at the supplied node
+  explicit depth_iterator_baset(const exprt &root)
+  {
+    this->downcast().push_expr(root);
+  }
+  typedef typename std::vector<exprt>::const_iterator operands_iteratort;
+
+  struct expr_statet final
+  {
+    expr_statet(
+      const exprt &expr,
+      operands_iteratort it,
+      operands_iteratort end):
+      expr(expr), it(it), end(end) { }
+    expr_statet(const expr_statet &other):
+      expr(other.expr), it(other.it), end(other.end) { }
+    std::reference_wrapper<const exprt> expr;
+    // Cache to avoid unnecessary mutable operands() calls
+    std::vector<exprt> *operands=nullptr;
+    operands_iteratort it;
+    const operands_iteratort end;
+    bool operator==(const expr_statet &other) const
+    {
+      return it==other.it;
+    }
+  };
+  // Obtain non-const exprt reference. Breaks the tree as a side
+  // effect. To be called only if a non-const exprt was passed into
+  // the constructor Do not call on the end() iterator
+  exprt &mutate()
+  {
+    PRECONDITION(!m_stack.empty());
+    auto expr=std::ref(const_cast<exprt&>(m_stack.front().expr.get()));
+    auto old_stack=std::move(m_stack);
+    try
+    {
+      m_stack.reserve(old_stack.size());
+      for(const auto& state : old_stack)
+      {
+        auto &operands=state.operands
```

Comment:
When `.mutate()` is called on more than one node during the same iteration. Since it is the only way to obtain non-const `exprt` references, I'd say that might be a common use case.